ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjector.java

```diff
@@ -155,45 +161,130 @@ private void registerSchemas(
+          schema.value(),
+          kafkaTopic,
+          valueFormat,
+          valueSerdeFeatures,
+          config,
+          statementText,
+          registerIfSchemaExists,
+          getSRSubject(kafkaTopic, false),
+          false
+      );
+    }
+
+    // Do all santity checks before register anything
+    if (registerRawKey) {
+      // validate
+      sanityCheck(kvRawSchema.left, keyFormat, kafkaTopic, config, statementText, true);
+    }
+    if (registerRawValue) {
+      // validate
+      sanityCheck(kvRawSchema.right, valueFormat, kafkaTopic, config, statementText, false);
+    }
+    if (registerRawKey) {
+      registerRawSchema(
+          kvRawSchema.left,
+          kafkaTopic,
+          statementText,
+          getSRSubject(kafkaTopic, true),
+          true);
+    }
+    if (registerRawValue) {
+      registerRawSchema(
+          kvRawSchema.right,
+          kafkaTopic,
+          statementText,
+          getSRSubject(kafkaTopic, false),
+          false);
+    }
   }
 
-  private void registerSchema(
-      final List<? extends SimpleColumn> schema,
-      final String topic,
+  private static void sanityCheck(
+      final SchemaAndId schemaAndId,
       final FormatInfo formatInfo,
-      final SerdeFeatures serdeFeatures,
+      final String topic,
       final KsqlConfig config,
       final String statementText,
-      final boolean registerIfSchemaExists,
-      final String subject,
       final boolean isKey
   ) {
+    final String schemaIdPropStr =
+        isKey ? CommonCreateConfigs.KEY_SCHEMA_ID : CommonCreateConfigs.VALUE_SCHEMA_ID;
     final Format format = FormatFactory.of(formatInfo);
+    if (!canRegister(format, config, topic)) {
+      throw new KsqlStatementException(schemaIdPropStr + " is provided but format "
+          + format.name() + " doesn't support registering in Schema Registry",
+          statementText);
+    }
+
+    final SchemaTranslator translator = format.getSchemaTranslator(formatInfo.getProperties());
+    if (!translator.name().equals(schemaAndId.rawSchema.schemaType())) {
+      throw new KsqlStatementException(String.format(
+          "Format and fetched schema type using %s %d are different. Format: [%s], "
+              + "Fetched schema type: [%s].",
+          schemaIdPropStr, schemaAndId.id, format.name(), schemaAndId.rawSchema.schemaType()),
+          statementText);
+    }
+  }
+
+  private void registerRawSchema(
+      final SchemaAndId schemaAndId,
+      final String topic,
+      final String statementText,
+      final String subject,
+      final Boolean isKey
+  ) {
+    final int id;
+    try {
+      id = SchemaRegistryUtil.registerSchema(serviceContext.getSchemaRegistryClient(),
+          schemaAndId.rawSchema, topic, subject, isKey);
+    } catch (KsqlException e) {
+      throw new KsqlStatementException("Could not register schema for topic: " + e.getMessage(),
+          statementText, e);
+    }
+
+    if (id != schemaAndId.id) {
```

Comment:
One thing I'm about update here is to check if `ServiceContext` is sandboxservicecontext and make it pass validation because sand box sr client always return `123` which will fail this. Do you think it's a good way to bypass this?

Reply:
I'm now wondering, why do we have to register a schema that should have already be registered? If a schema_id, then we just need to fetch the scheme and use it, right?

I was going to say, what if you do this check inside the `SchemaRegistryUtil.registerSchema`, so it does not interfer with the sandbox, but I got wonder about the above question.

Reply:
Discussed offline. Fetched raw schema can be in any subject. Reregister here under the correct subject to have proper compatibility check in SR. Also during serialization time using schema id, serializer will look up the schema under correct subject.